termuic/termuiConsole.go

```diff
@@ -0,0 +1,214 @@
+package termuic
+
+import (
+	"os"
+	"os/signal"
+	"runtime"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	ui "github.com/gizak/termui/v3"
+)
+
+//refreshInterval is used for a ticker that refresh termui console at a specific interval
+const refreshInterval = time.Second
+
+// TermuiConsole data where is store data from handler
+type TermuiConsole struct {
+	termuiConsoleMetrics *sync.Map
+	logLines             []string
+	mutLogLineWrite      sync.RWMutex
+	numLogLines          int
+	termuiConsoleWidgets *termuiConsoleGrid
+	grid                 *ui.Grid
+}
+
+//NewTermuiConsole method is used to return a new TermuiConsole structure
+func NewTermuiConsole(metricData *sync.Map) *TermuiConsole {
+	tc := TermuiConsole{
+		termuiConsoleMetrics: metricData,
+		numLogLines:          10,
+	}
+
+	return &tc
+}
+
+func (tc *TermuiConsole) Write(p []byte) (n int, err error) {
+	go func() {
+		logLine := string(p)
+		logLine = strings.Replace(logLine, "\n", "", len(logLine))
+		logLine = strings.Replace(logLine, "\r", "", len(logLine))
+
+		tc.mutLogLineWrite.Lock()
+		if len(tc.logLines) >= tc.numLogLines {
+			tc.logLines = tc.logLines[1:tc.numLogLines]
+		}
+		if logLine != "" {
+			tc.logLines = append(tc.logLines, logLine)
+		}
+		tc.mutLogLineWrite.Unlock()
+	}()
+
+	return len(p), nil
+}
+
+// Start method - will start termui console
+func (tc *TermuiConsole) Start() error {
+
+	if err := ui.Init(); err != nil {
+		return err
+	}
+
+	go func() {
+		defer ui.Close()
+
+		tc.eventLoop()
+	}()
+
+	return nil
+}
+
+func (tc *TermuiConsole) eventLoop() {
+
+	tc.termuiConsoleWidgets = NewtermuiConsoleGrid()
+
+	time.Sleep(1 * time.Second)
+
+	tc.grid = tc.configConsoleWithNormalTermuiWidgets()
+
+	uiEvents := ui.PollEvents()
+	// handles kill signal sent to gotop
+	sigTerm := make(chan os.Signal, 2)
+	signal.Notify(sigTerm, os.Interrupt, syscall.SIGTERM)
+
+	drawTicker := time.NewTicker(refreshInterval).C
+
+	for {
+		select {
+		case <-drawTicker:
+			tc.refreshDataForConsole()
+
+		case <-sigTerm:
+			ui.Clear()
+
+			return
+		case e := <-uiEvents:
+			tc.processUiEvents(e)
+		}
+	}
+}
+
+func (tc *TermuiConsole) processUiEvents(e ui.Event) {
+	switch e.ID {
+	case "<Resize>":
+		payload := e.Payload.(ui.Resize)
+		tc.grid.SetRect(0, 0, payload.Width, payload.Height)
+		ui.Clear()
+		ui.Render(tc.grid)
+	case "q":
+		ui.Clear()
+		tc.grid = tc.changeConsoleDisplay()
+
+	case "<C-c>":
+		ui.Close()
+		if p, err := os.FindProcess(os.Getpid()); err != nil {
+			return
+		} else {
+			if runtime.GOOS == "windows" {
+				_ = p.Kill()
+			} else {
+				_ = p.Signal(syscall.SIGINT)
+			}
+		}
+		return
+	}
+}
+
+func (tc *TermuiConsole) refreshDataForConsole() {
+	if tc.numLogLines == 10 {
+		tc.prepareDataNormalWidgets()
+	}
+	tc.termuiConsoleWidgets.PrepareListWithLogsForDisplay(tc.logLines)
```

Comment:
Should this be on the else?

Reply:
this code was changed now I m using a interface for display console with normal widgets or with a big log.